src/metabase/feature_extraction/timeseries.clj

```diff
@@ -0,0 +1,174 @@
+(ns metabase.feature-extraction.timeseries
+  "Timeseries analysis and utilities."
+  (:require [bigml.histogram.core :as h.impl]
+            [clj-time
+             [coerce :as t.coerce]
+             [core :as t]
+             [periodic :as t.periodic]]
+            [kixi.stats
+             [core :refer [somef] :as stast]
+             [math :as k.math]]
+            [metabase.feature-extraction
+             [histogram :as h]
+             [math :as math]]
+            [redux.core :as redux])
+  (:import (com.github.brandtg.stl StlDecomposition StlResult StlConfig)))
+
+(def ^{:arglists '([t])} to-double
+  "Coerce `DateTime` to `Double`."
+  (comp double t.coerce/to-long))
+
+(def ^{:arglists '([t])} from-double
+  "Coerce `Double` into a `DateTime`."
+  (somef (comp t.coerce/from-long long)))
+
+(defn fill-timeseries
+  "Given a coll of `[DateTime, Num]` pairs evenly spaced `step` apart, fill
+   missing points with 0."
+  [resolution ts]
+  (let [[step rounder] (case resolution
+                         :month   [(t/months 1) t/month]
+                         :quarter [(t/months 3) t/month]
+                         :year    [(t/years 1) t/year]
+                         :week    [(t/weeks 1) t/day]
+                         :day     [(t/days 1) t/day]
+                         :hour    [(t/hours 1) t/day]
+                         :minute  [(t/minutes 1) t/minute])
+        ts             (for [[x y] ts]
+                         [(-> x from-double (t/floor rounder)) y])
+        ts-index       (into {} ts)]
+    (into []
+      (comp (take-while (partial (complement t/before?) (-> ts last first)))
+            (map (fn [t]
+                   [(to-double t) (or (ts-index t) 0)])))
+      (some-> ts
+              ffirst
+              (t.periodic/periodic-seq step)))))
+
+(def period-length
+  "What is the period for a given time resolution."
+  {:hour         24
+   :minute       60
+   :month        12
+   :week         52
+   :quarter      4
+   :day          365})
+
+(def ^:private stl-setters
+  {:inner-loop-passes           (memfn ^StlConfig setNumberOfInnerLoopPasses n)
+   :robustness-iterations       (memfn ^StlConfig setNumberOfRobustnessIterations n)
+   :trend-bandwidth             (memfn ^StlConfig setTrendComponentBandwidth bw)
+   :seasonal-bandwidth          (memfn ^StlConfig setSeasonalComponentBandwidth bw)
+   :loess-robustness-iterations (memfn ^StlConfig setLoessRobustnessIterations n)
+   :periodic?                   (memfn ^StlConfig setPeriodic periodic?)})
+
+(defn decompose
+  "Decompose given timeseries with expected periodicty `period` into trend,
+   seasonal component, and residual.
+   `period` can be one of `:hour`, `:day`, `:week`, `:quarter` `:minute`,
+   `:month`, or `:year`.
+   https://www.wessa.net/download/stl.pdf"
+  ([period ts]
+   (decompose period {} ts))
+  ([period opts ts]
+   (when-let [period (period-length period)]
```

Comment:
I'm having difficulty matching up the tests (and the boolean return value) with the code here. It looks like it's config driving the setup of the StlConfig objects, but what happens after that?

Reply:
It does the actual decomposition in the cleanup airity of the transducer. Clearly too cute. Refactored.